fuse_ll/src/fuse/mount.rs

```diff
@@ -10,21 +11,157 @@ use std::path::Path;
 
 use param::*;
 
+pub struct MountOption {
+    pub name: String,
+    pub parser: fn(&mut FuseMountArgs, &MountOption, &str),
+    pub regex: Regex,
+    #[cfg(target_os = "linux")]
+    pub flag: Option<u64>,
+    #[cfg(target_os = "macos")]
+    pub flag: Option<i32>,
+    #[cfg(target_os = "macos")]
+    pub fuse_flag: Option<u64>,
+}
+
+pub fn get_all_options() -> String {
+    let options: Vec<_> = get_mount_options().iter().map(|x| x.name.clone()).collect();
+    options.join(",")
+}
+
+pub fn options_validator(option: String) -> Result<(), String> {
+    let ops: Vec<_> = option.split(',').collect();
+    let ret = ops
+        .iter()
+        .all(|&op| get_mount_options().iter().any(|x| x.regex.is_match(&op)));
+    if ret {
+        Ok(())
+    } else {
+        Err(format!(
+            "Invalid option \"{}\", valid options: {}",
+            option,
+            get_all_options()
+        ))
+    }
+}
+
 #[cfg(target_os = "linux")]
 mod param {
     // https://github.com/torvalds/linux/blob/master/include/uapi/linux/mount.h#L11
     // TODO: use mount flags from libc
-    // pub const MS_RDONLY: u64 = 1; // Mount read-only
+    pub const MS_RDONLY: u64 = 1; // Mount read-only
     pub const MS_NOSUID: u64 = 2; // Ignore suid and sgid bits
     pub const MS_NODEV: u64 = 4; // Disallow access to device special files
     pub const MNT_FORCE: i32 = 1; // Force un-mount
+
+    use super::MountOption;
+    use regex::Regex;
+    fn add_option(options: &Option<String>, option: &str) -> Option<String> {
+        match options {
+            None => Some(String::from(option)),
+            Some(s) => Some(s.to_owned() + "," + option),
+        }
+    }
+
+    pub fn get_mount_options() -> Vec<MountOption> {
+        fn parse_flag(args: &mut FuseMountArgs, mount_option: &MountOption, _options: &str) {
+            if let Some(flag) = mount_option.flag {
+                args.flags |= flag;
+            }
+        }
+
+        fn parse_allow_other(args: &mut FuseMountArgs, _mount_option: &MountOption, option: &str) {
+            args.allow_other = 1;
+            args.kernel_opts = add_option(&args.kernel_opts, option);
+        }
+
+        fn parse_fsname(args: &mut FuseMountArgs, _mount_option: &MountOption, option: &str) {
+            let name = String::from(option.split('=').last().unwrap()); //Safe to use unwrap here, becuase option is always valid.
+            args.fsname = Some(name);
+            args.fusermount_opts = add_option(&args.fusermount_opts, option);
+        }
+        // Safe to use unwrap here, because expression is literal string and always valid.
+        let ro_regex = Regex::new("^ro$").unwrap();
+        let rw_regex = Regex::new("^rw$").unwrap();
+        let allow_other_regex = Regex::new("^allow_other$").unwrap();
+        let fsname_regex = Regex::new("^fsname=.+$").unwrap();
```

Comment:
will white space be included in the regex? ".+" will include all possible characters.

Reply:
Will fix.